File: app/src/main/java/com/hades/mvvmdatabindinglivedata/data/source/api/service/ServiceGenerator.java
package com.hades.mvvmdatabindinglivedata.data.source.api.service;

import com.hades.mvvmdatabindinglivedata.data.source.api.service.GithubService;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by Hades on 9/10/2018.
 */
public class ServiceGenerator {

    private static final String BASE_URL = "https://api.github.com/";
    private static Retrofit sInstance;

    public static GithubService createService() {

        HttpLoggingInterceptor httpLoggingInterceptor =
                new HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient okHttpClient =
                new OkHttpClient.Builder().addInterceptor(httpLoggingInterceptor).build();

        if (sInstance == null) {
            sInstance = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .client(okHttpClient)
                    .build();
        }
        return sInstance.create(GithubService.class);
    }
}

File: app/src/main/java/com/hades/kotlintrainning/data/api/ServiceGenerator.kt
package com.hades.kotlintrainning.data.api

import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.Retrofit
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor

object ServiceGenerator {

    private const val BASE_URL = "https://api.themoviedb.org/"
    private var sRetrofit: Retrofit? = null

    private val httpLoggingInterceptor = HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY)

    private val okHttpClient = OkHttpClient.Builder().addInterceptor(httpLoggingInterceptor).build()

    fun createService(): ApiService {

        if (sRetrofit == null) {
            sRetrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .client(okHttpClient)
                .build()
        }
        return sRetrofit!!.create(ApiService::class.java)
    }
}